### ISSUE TITLE
Can we represent indices in the k parameter?

### ISSUE THREAD

**Issue by adamgundry** [well-typed/optics#232]
See https://www.reddit.com/r/haskell/comments/cyo4o2/welltyped_announcing_the_optics_library/eytbj1t/

**Comment by phadej** [comment]
Being indexed is orthogonal to the kind of the optic. Even `Iso` can be indexed, it’s just not so useful.

**Comment by adamgundry** [comment]
Indeed. It's not clear to me that this rules out packing both the index and the kind into a single parameter though. Perhaps we can investigate at MuniHac. :smile:

**Comment by phadej** [comment]
@adamgundry it would make interface more complicated: as `A_Traversal` and `An_IndexedTraversal '[]` is the same thing. Or you could do `An_IndexedTraversal i '[]` (i.e. require at least one index of indexed version). Either way the `Join` size would be squared (which would also need to compute indices) and similar trickiness with `Is`.

There is no experience in teaching indexed optics to beginners in general. My gut feeling is that orthogonal constructs are simpler, even there is an additional type-argument in `Optic` which is `NoIx` for time being.

**Comment by adamgundry** [comment]
I'd wondered whether it works to use a single wrapper type `Indexed i k` where `k` is an optic kind, so
```hs
type IndexedTraversal i s t a b = Optic (Indexed i A_Traversal) s t a b

Join (Indexed i k) l = Indexed i (Join k l)
Join k (Indexed l) = Indexed l (Join k l) -- provided k not Indexed
```
Then I guess the tricky part is the type of composition, because we can no longer use `Is k (Join k l)` but have to attach the indices from the other optic. Perhaps we could make it work using something like this, for suitable definitions of type families `IndicesAfter` and `IndicesBefore`:
```hs
(%)
  :: (Is (IndicesAfter k l) (Join k l), Is (IndicesBefore l k) (Join k l)) =>
     Optic k s t u v
     -> Optic l u v a b -> Optic (Join k l) s t a b
```
or perhaps we can (re)introduce a typeclass constraint `CanCompose k l`. But it isn't clear how the implementation should work, and I agree this is feeling rather non-orthogonal, so I'm inclined to just give up the idea.
